open-craft/openedx-edit-links
# Implement "Edit on Git" link for each section

### Description
The first feature functional PR to the edit_links Open edX extension.

* Adds a model with admin view to configure course specfiic repositories
  and the base_url to the edit interfaces of the source code hosting
  services that hold the course information.
* Implements the pipeline for
  `org.openedx.learning.vertical_block_child.render.started.v1` event, that can
  generate the "Edit this on <some-service>" links to the XBlocks based
  on the XBlock IDs

**JIRA:**

- Full description of the issue: https://gitlab.com/mooc-floss/mooc-floss/-/issues/121

**Dependencies:** 
- edx-platform - https://github.com/openedx/edx-platform/pull/30773
- openedx-filters - https://github.com/openedx/openedx-filters/pull/38

**Installation instructions:**

This assumes you are testing this on a devstack.

1. Checkout the [edx-platform PR](https://github.com/openedx/edx-platform/pull/30773) branch. Note: add the **open-craft** remote if you don't have it already.
    ```
    git checkout open-craft/tecoholic/add-filter-hook-to-vertical-block-rendering
    ```
2. Clone the edx-filters repository to the `src` folder and checkout to the [corresponding PR](https://github.com/openedx/openedx-filters/pull/38) branch 
    ```
    cd <your-devstack-location>/src
    git clone https://github.com/open-craft/openedx-filters.git
    # if you already have openedx repo checkout, add the open-craft remote and fetch the PR branch
    git checkout tecoholic/add-vertical-block-child-render-started-filter
    ```
3. Clone the repository to your `src` directory in your devstack setup and checkout to the PR branch
    ```
    git clone https://github.com/open-craft/openedx-edit-links.git
    git checkout tecoholic/BB-6206-edit-links-poc
    ```
5. Drop into LMS shell and install the openedx-filters extension and the openedx-edit-links extension
    ```
    make lms-shell
    pip install -e /edx/src/openedx-filters
    pip install -e /edx/src/openedx-edit-links
    ```
6. Add the following configuration to `lms/envs/private.py` of `edx-platform`
    ```python
    OPEN_EDX_FILTERS_CONFIG = {
            "org.openedx.learning.vertical_block_child.render.started.v1": {
                "fail_silently": False,
                "pipeline": [
                    "edit_links.pipeline.AddEditLink"
                ]
            }
        }
    EDIT_LINKS_PLUGIN_GIT_REPOS = {
    }
    EDIT_LINKS_PLUGIN_GIT_EDIT_LABEL = "Gitlab"
    ```
7. Restart the LMS `make lms-restart`

**Testing instructions:**

One the extension is installed and configured as above, we can test the extension using the MOOC FLOSS course content.

1. Download the `course` folder from https://gitlab.com/mooc-floss/mooc-floss/-/tree/master/course as a `.tar.gz` file. 
![poc-download-course](https://user-images.githubusercontent.com/383862/175516855-b6fafe27-9e57-4f98-855d-00af647fa0c3.png)
2. Create a new course in studio and go to **Tools** > **Import** and upload the `.tar.gz` file just downloaded.
3. If you face a "problem-builder" related error during the import, install the [Problem Builder XBlock](https://github.com/open-craft/problem-builder) in both studio and lms containers and then try the import process again.
    ```
    make lms-shell
    pip install xblock-problem-builder
    make studio-shell
    pip install xblock-problem-builder
    ```
4. Ensure the course is successfully imported and the units are loading in the LMS. Make a note of the course ID eg., `course-v1:FLOSS+OSS101+2022_T4`
5. Edit the `EDIT_LINKS_PLUGIN_GIT_REPOS` value in `private.py` to look like the following (use your course's ID)
    ```
    EDIT_LINKS_PLUGIN_GIT_REPOS = {
        "course-v1:FLOSS+OSS101+2022_T4": "https://gitlab.com/-/ide/project/mooc-floss/mooc-floss/edit/master/-/course/html/"
    }
    ```
6. Now re-open the course in LMS, each section of the unit should have a **Edit in Gitlab** link on the top. 
![Screenshot with Edit Links](https://user-images.githubusercontent.com/383862/180652172-723cce6c-96ce-4003-a313-549fb2a3d7ed.png)
7. Clicking the link should take the user directly to the Gitlab IDE with the contents of the unit available as editable HTML.

**Reviewers:**
- [x] @navinkarkera

cc: @antoviaque @pomegranited 

**Merge checklist:**
- [ ] openedx-filters version changed from custom branch to release version

**Post merge:**
      finished.

**Author concerns:**

We use [a protected method of the XBlock](https://github.com/open-craft/openedx-edit-links/pull/1/files#diff-45e67145a1f2842edfcd1f05ceec3e4213c8848c66a05c15ceedd699c82b21eeR56) to keep the HTML of the Edit Link from raising error due to XBlock's render function auto-saving any changes.
